File: internal/store/fts.go
package store

import (
	"context"
	"fmt"
	"sort"
	"strings"
	"unicode/utf8"
)

// Sub-type constants used across the store package.
const (
	SubTypeGeneral  = "general"
	SubTypeDecision = "decision"
	SubTypePattern  = "pattern"
	SubTypeRule     = "rule"
)

// SearchKnowledgeFTS searches knowledge using FTS5 with tag alias expansion.
// Falls back to keyword LIKE search if FTS5 fails.
func (s *Store) SearchKnowledgeFTS(ctx context.Context, query string, limit int) ([]KnowledgeRow, error) {
	if limit <= 0 {
		limit = 10
	}
	query = strings.TrimSpace(query)
	if query == "" {
		return s.SearchKnowledgeKeyword(ctx, "", limit)
	}

	// Expand query terms with aliases.
	words := strings.Fields(query)
	expanded, err := s.ExpandAliases(ctx, words)
	if err != nil {
		expanded = words
	}

	// Build FTS5 OR query.
	var ftsTerms []string
	for _, w := range expanded {
		w = sanitizeFTSTerm(w)
		if w != "" {
			ftsTerms = append(ftsTerms, `"`+w+`"`)
		}
	}
	if len(ftsTerms) == 0 {
		return s.SearchKnowledgeKeyword(ctx, query, limit)
	}
	ftsQuery := strings.Join(ftsTerms, " OR ")

	docs, err := s.searchFTSKnowledge(ctx, ftsQuery, limit)
	if err != nil {
		return s.SearchKnowledgeKeyword(ctx, query, limit)
	}

	// Supplement with fuzzy matching if too few results.
	if len(docs) < limit {
		fuzzyDocs := s.fuzzySearchKnowledge(ctx, words, limit-len(docs), docs)
		docs = append(docs, fuzzyDocs...)
	}

	return docs, nil
}

// searchFTSKnowledge runs FTS5 search on knowledge_fts.
func (s *Store) searchFTSKnowledge(ctx context.Context, ftsQuery string, limit int) ([]KnowledgeRow, error) {
	sqlQuery := `SELECT k.id, k.file_path, k.content_hash, k.title, k.content, k.sub_type,
		k.project_remote, k.project_path, k.project_name, k.branch,
		k.created_at, k.updated_at, k.hit_count, k.last_accessed, k.enabled,
		bm25(knowledge_fts, 3.0, 1.0, 1.0) AS rank
	FROM knowledge_fts f
	JOIN knowledge_index k ON k.id = f.rowid
	WHERE knowledge_fts MATCH ? AND k.enabled = 1
	ORDER BY rank
	LIMIT ?`

	rows, err := s.db.QueryContext(ctx, sqlQuery, ftsQuery, limit)
	if err != nil {
		return nil, err
	}
	defer rows.Close()

	var docs []KnowledgeRow
	for rows.Next() {
		var r KnowledgeRow
		var rank float64
		if err := rows.Scan(
			&r.ID, &r.FilePath, &r.ContentHash, &r.Title, &r.Content, &r.SubType,
			&r.ProjectRemote, &r.ProjectPath, &r.ProjectName, &r.Branch,
			&r.CreatedAt, &r.UpdatedAt, &r.HitCount, &r.LastAccessed, &r.Enabled,
			&rank,
		); err != nil {
			continue
		}
		docs = append(docs, r)
	}
	return docs, rows.Err()
}

// fuzzySearchKnowledge scans knowledge and returns those matching via fuzzy distance.
func (s *Store) fuzzySearchKnowledge(ctx context.Context, queryWords []string, limit int, exclude []KnowledgeRow) []KnowledgeRow {
	if limit <= 0 {
		return nil
	}
	excludeIDs := make(map[int64]bool, len(exclude))
	for _, d := range exclude {
		excludeIDs[d.ID] = true
	}

	rows, err := s.db.QueryContext(ctx,
		`SELECT id, file_path, content_hash, title, content, sub_type,
		        project_remote, project_path, project_name, branch,
		        created_at, updated_at, hit_count, last_accessed, enabled
		 FROM knowledge_index WHERE enabled = 1 LIMIT 500`)
	if err != nil {
		return nil
	}
	defer rows.Close()

	var docs []KnowledgeRow
	for rows.Next() {
		var r KnowledgeRow
		if err := rows.Scan(
			&r.ID, &r.FilePath, &r.ContentHash, &r.Title, &r.Content, &r.SubType,
			&r.ProjectRemote, &r.ProjectPath, &r.ProjectName, &r.Branch,
			&r.CreatedAt, &r.UpdatedAt, &r.HitCount, &r.LastAccessed, &r.Enabled,
		); err != nil {
			continue
		}

		if excludeIDs[r.ID] {
			continue
		}

		// Fuzzy match against title words.
		targetWords := strings.Fields(strings.ToLower(r.Title))
		for _, qw := range queryWords {
			matched := false
			for _, tw := range targetWords {
				if FuzzyMatch(qw, tw) {
					matched = true
					break
				}
			}
			if matched {
				docs = append(docs, r)
				if len(docs) >= limit {
					return docs
				}
				break
			}
		}
	}
	return docs
}

// KnowledgeConflict represents a pair of potentially contradictory or duplicate knowledge entries.
type KnowledgeConflict struct {
	A          KnowledgeRow
	B          KnowledgeRow
	Similarity float64
	Type       string // "potential_duplicate" or "potential_contradiction"
}

// DetectKnowledgeConflicts finds pairs of knowledge entries with high cosine similarity.
func (s *Store) DetectKnowledgeConflicts(ctx context.Context, threshold float64) ([]KnowledgeConflict, error) {
	if threshold <= 0 {
		threshold = 0.70
	}

	rows, err := s.db.QueryContext(ctx,
		`SELECT e.source_id, e.vector FROM embeddings e
		 JOIN knowledge_index k ON k.id = e.source_id
		 WHERE e.source = 'knowledge' AND k.enabled = 1
		 LIMIT 1000`)
	if err != nil {
		return nil, fmt.Errorf("store: detect conflicts query: %w", err)
	}
	defer rows.Close()

	type embeddedDoc struct {
		id  int64
		vec []float32
	}
	var docs []embeddedDoc
	for rows.Next() {
		var ed embeddedDoc
		var blob []byte
		if err := rows.Scan(&ed.id, &blob); err != nil {
			continue
		}
		ed.vec = deserializeFloat32(blob)
		docs = append(docs, ed)
	}

	// Pairwise cosine similarity.
	var conflicts []KnowledgeConflict
	for i := 0; i < len(docs); i++ {
		for j := i + 1; j < len(docs); j++ {
			if len(docs[i].vec) != len(docs[j].vec) {
				continue
			}
			sim := cosineSimilarity(docs[i].vec, docs[j].vec)
			if sim >= threshold {
				conflicts = append(conflicts, KnowledgeConflict{
					A:          KnowledgeRow{ID: docs[i].id},
					B:          KnowledgeRow{ID: docs[j].id},
					Similarity: sim,
				})
			}
		}
	}

	// Hydrate and classify.
	if len(conflicts) > 0 {
		var allIDs []int64
		for _, c := range conflicts {
			allIDs = append(allIDs, c.A.ID, c.B.ID)
		}
		hydrated, err := s.GetKnowledgeByIDs(ctx, allIDs)
		if err == nil {
			docMap := make(map[int64]KnowledgeRow, len(hydrated))
			for _, d := range hydrated {
				docMap[d.ID] = d
			}
			for i := range conflicts {
				if d, ok := docMap[conflicts[i].A.ID]; ok {
					conflicts[i].A = d
				}
				if d, ok := docMap[conflicts[i].B.ID]; ok {
					conflicts[i].B = d
				}
				conflicts[i].Type = classifyConflict(conflicts[i].A.Content, conflicts[i].B.Content)
			}
		}
	}

	sort.Slice(conflicts, func(i, j int) bool {
		return conflicts[i].Similarity > conflicts[j].Similarity
	})

	return conflicts, nil
}

// ExpandAliases expands search terms using the tag_aliases table.
func (s *Store) ExpandAliases(ctx context.Context, terms []string) ([]string, error) {
	if len(terms) == 0 {
		return nil, nil
	}

	seen := make(map[string]bool, len(terms)*2)
	for _, t := range terms {
		seen[strings.ToLower(t)] = true
	}

	for _, t := range terms {
		lower := strings.ToLower(t)

		rows, err := s.db.QueryContext(ctx,
			`SELECT alias FROM tag_aliases WHERE LOWER(tag) = ?`, lower)
		if err != nil {
			return nil, fmt.Errorf("store: expand aliases: %w", err)
		}
		for rows.Next() {
			var alias string
			if err := rows.Scan(&alias); err != nil {
				continue
			}
			seen[strings.ToLower(alias)] = true
		}
		rows.Close()

		rows, err = s.db.QueryContext(ctx,
			`SELECT tag FROM tag_aliases WHERE LOWER(alias) = ?`, lower)
		if err != nil {
			return nil, fmt.Errorf("store: expand aliases: %w", err)
		}
		for rows.Next() {
			var tag string
			if err := rows.Scan(&tag); err != nil {
				continue
			}
			seen[strings.ToLower(tag)] = true
		}
		rows.Close()
	}

	result := make([]string, 0, len(seen))
	for t := range seen {
		result = append(result, t)
	}
	return result, nil
}

// SubTypeHalfLife returns the recency decay half-life in days for a sub_type.
func SubTypeHalfLife(subType string) float64 {
	switch subType {
	case "assumption":
		return 30.0
	case "inference":
		return 45.0
	case SubTypeGeneral:
		return 60.0
	case SubTypePattern:
		return 90.0
	case SubTypeDecision:
		return 90.0
	case SubTypeRule:
		return 120.0
	default:
		return 60.0
	}
}

// SubTypeBoost returns a relevance multiplier based on sub_type.
func SubTypeBoost(subType string) float64 {
	switch subType {
	case SubTypeRule:
		return 2.0
	case SubTypeDecision:
		return 1.5
	case SubTypePattern:
		return 1.3
	default:
		return 1.0
	}
}

// buildFTSQuery converts a query into FTS5 syntax (AND-joined quoted terms).
func buildFTSQuery(query string) string {
	words := strings.Fields(query)
	var terms []string
	for _, w := range words {
		w = sanitizeFTSTerm(w)
		if w == "" {
			continue
		}
		terms = append(terms, `"`+w+`"`)
	}
	return strings.Join(terms, " AND ")
}

// sanitizeFTSTerm removes FTS5 special characters.
func sanitizeFTSTerm(term string) string {
	var b strings.Builder
	for _, r := range term {
		if r == '"' || r == '*' || r == '^' || r == '{' || r == '}' {
			continue
		}
		b.WriteRune(r)
	}
	return b.String()
}

// classifyConflict determines if a high-similarity pair is a duplicate or contradiction.
func classifyConflict(contentA, contentB string) string {
	lowerA := strings.ToLower(contentA)
	lowerB := strings.ToLower(contentB)

	for _, pair := range contradictionPairs {
		aHas0 := strings.Contains(lowerA, pair[0])
		aHas1 := strings.Contains(lowerA, pair[1])
		bHas0 := strings.Contains(lowerB, pair[0])
		bHas1 := strings.Contains(lowerB, pair[1])

		if (aHas0 && bHas1 && !aHas1) || (aHas1 && bHas0 && !bHas1) {
			return "potential_contradiction"
		}
	}
	return "potential_duplicate"
}

var contradictionPairs = [][2]string{
	{"always", "never"},
	{"must", "must not"},
	{"use", "avoid"},
	{"enable", "disable"},
	{"allow", "deny"},
	{"required", "optional"},
	{"do", "don't"},
	{"add", "remove"},
	{"include", "exclude"},
}

// Levenshtein computes the Levenshtein distance between two strings.
func Levenshtein(a, b string) int {
	ra := []rune(a)
	rb := []rune(b)
	la := len(ra)
	lb := len(rb)

	if la == 0 {
		return lb
	}
	if lb == 0 {
		return la
	}

	prev := make([]int, lb+1)
	for j := range prev {
		prev[j] = j
	}

	for i := 1; i <= la; i++ {
		curr := make([]int, lb+1)
		curr[0] = i
		for j := 1; j <= lb; j++ {
			cost := 1
			if ra[i-1] == rb[j-1] {
				cost = 0
			}
			curr[j] = min(curr[j-1]+1, prev[j]+1, prev[j-1]+cost)
		}
		prev = curr
	}
	return prev[lb]
}

// FuzzyMatch returns true if query is within acceptable edit distance of target.
func FuzzyMatch(query, target string) bool {
	qLen := utf8.RuneCountInString(query)
	if qLen < 3 {
		return false
	}
	maxDist := min(2, qLen/3)
	if maxDist == 0 {
		maxDist = 1
	}
	return Levenshtein(strings.ToLower(query), strings.ToLower(target)) <= maxDist
}

File: src/store/fts.ts
import type { KnowledgeConflict, KnowledgeRowV1 as KnowledgeRow } from "../types.js";
import type { Store } from "./index.js";
import type { RawKnowledgeRow } from "./knowledge.js";
import { getKnowledgeByIDs, mapRow, searchKnowledgeKeyword } from "./knowledge.js";
import { pairwiseSimilarity } from "./vectors.js";

export function subTypeHalfLife(subType: string): number {
	switch (subType) {
		case "assumption":
			return 30;
		case "inference":
			return 45;
		case "snapshot":
			return 30;
		case "pattern":
			return 90;
		case "decision":
			return 90;
		case "rule":
			return 120;
		default:
			return 60;
	}
}

export function subTypeBoost(subType: string): number {
	switch (subType) {
		case "rule":
			return 2.0;
		case "decision":
			return 1.5;
		case "pattern":
			return 1.3;
		default:
			return 1.0;
	}
}

export function searchKnowledgeFTS(store: Store, query: string, limit: number): KnowledgeRow[] {
	if (limit <= 0) limit = 10;
	query = query.trim();
	if (!query) return searchKnowledgeKeyword(store, "", limit);

	const words = query.split(/\s+/);
	let expanded: string[];
	try {
		expanded = expandAliases(store, words);
	} catch {
		expanded = words;
	}

	const ftsTerms: string[] = [];
	for (const w of expanded) {
		const sanitized = sanitizeFTSTerm(w);
		if (sanitized) ftsTerms.push(`"${sanitized}"`);
	}
	if (ftsTerms.length === 0) {
		return searchKnowledgeKeyword(store, query, limit);
	}
	const ftsQuery = ftsTerms.join(" OR ");

	let docs: KnowledgeRow[];
	try {
		docs = searchFTSKnowledge(store, ftsQuery, limit);
	} catch {
		return searchKnowledgeKeyword(store, query, limit);
	}

	if (docs.length < limit) {
		const fuzzyDocs = fuzzySearchKnowledge(store, words, limit - docs.length, docs);
		docs.push(...fuzzyDocs);
	}

	return docs;
}

function searchFTSKnowledge(store: Store, ftsQuery: string, limit: number): KnowledgeRow[] {
	const rows = store.db
		.prepare(`
    SELECT k.id, k.project_id, k.file_path, k.content_hash, k.title, k.content, k.sub_type,
      k.branch, k.created_at, k.updated_at, k.hit_count, k.last_accessed, k.enabled,
      bm25(knowledge_fts, 3.0, 1.0, 1.0) AS rank
    FROM knowledge_fts f
    JOIN knowledge_index k ON k.id = f.rowid
    WHERE knowledge_fts MATCH ? AND k.enabled = 1
    ORDER BY rank
    LIMIT ?
  `)
		.all(ftsQuery, limit) as Array<RawKnowledgeRow & { rank: number }>;

	return rows.map(mapRow);
}

// --- Unified search (knowledge + spec) ---

export type UnifiedSource = "knowledge" | "spec";

export interface UnifiedSearchResult {
	id: number;
	source: UnifiedSource;
	title: string;
	content: string;
	projectId: string;
	projectName: string;
	score: number;
	// knowledge-specific
	subType?: string;
	hitCount?: number;
	// spec-specific
	slug?: string;
	fileName?: string;
	specStatus?: string;
}

export function searchUnified(
	store: Store,
	query: string,
	options?: {
		sources?: UnifiedSource[];
		projectId?: string;
		limit?: number;
	},
): UnifiedSearchResult[] {
	const limit = options?.limit ?? 20;
	const sources = options?.sources ?? ["knowledge", "spec"];
	const projectId = options?.projectId;

	const q = query.trim();
	if (!q) return [];

	const words = q.split(/\s+/);
	let expanded: string[];
	try {
		expanded = expandAliases(store, words);
	} catch {
		expanded = words;
	}
	const ftsTerms: string[] = [];
	for (const w of expanded) {
		const sanitized = sanitizeFTSTerm(w);
		if (sanitized) ftsTerms.push(`"${sanitized}"`);
	}
	if (ftsTerms.length === 0) return [];
	const ftsQuery = ftsTerms.join(" OR ");

	const results: UnifiedSearchResult[] = [];

	if (sources.includes("knowledge")) {
		const projectFilter = projectId ? "AND k.project_id = ?" : "";
		const params: unknown[] = [ftsQuery];
		if (projectId) params.push(projectId);
		params.push(limit);

		try {
			const rows = store.db
				.prepare(`
        SELECT k.id, k.project_id, k.title, k.content, k.sub_type, k.hit_count,
               COALESCE(p.name, '') as project_name,
               bm25(knowledge_fts, 3.0, 1.0, 1.0) AS rank
        FROM knowledge_fts f
        JOIN knowledge_index k ON k.id = f.rowid
        LEFT JOIN projects p ON p.id = k.project_id
        WHERE knowledge_fts MATCH ? AND k.enabled = 1 ${projectFilter}
        ORDER BY rank
        LIMIT ?
      `)
				.all(...params) as Array<{
				id: number; project_id: string; title: string; content: string;
				sub_type: string; hit_count: number; project_name: string; rank: number;
			}>;

			for (const r of rows) {
				results.push({
					id: r.id, source: "knowledge", title: r.title, content: r.content,
					projectId: r.project_id, projectName: r.project_name,
					score: -r.rank, // BM25 rank is negative (lower = better)
					subType: r.sub_type, hitCount: r.hit_count,
				});
			}
		} catch (e) { process.stderr.write(`alfred: FTS knowledge search error: ${e}\n`); }
	}

	if (sources.includes("spec")) {
		const projectFilter = projectId ? "AND s.project_id = ?" : "";
		const params: unknown[] = [ftsQuery];
		if (projectId) params.push(projectId);
		params.push(limit);

		try {
			const rows = store.db
				.prepare(`
        SELECT s.id, s.project_id, s.title, s.content, s.slug, s.file_name, s.status,
               COALESCE(p.name, '') as project_name,
               bm25(spec_fts, 3.0, 1.0, 1.0) AS rank
        FROM spec_fts f
        JOIN spec_index s ON s.id = f.rowid
        LEFT JOIN projects p ON p.id = s.project_id
        WHERE spec_fts MATCH ? ${projectFilter}
        ORDER BY rank
        LIMIT ?
      `)
				.all(...params) as Array<{
				id: number; project_id: string; title: string; content: string;
				slug: string; file_name: string; status: string; project_name: string; rank: number;
			}>;

			for (const r of rows) {
				results.push({
					id: r.id, source: "spec", title: r.title, content: r.content,
					projectId: r.project_id, projectName: r.project_name,
					score: -r.rank,
					slug: r.slug, fileName: r.file_name, specStatus: r.status,
				});
			}
		} catch (e) { process.stderr.write(`alfred: FTS spec search error: ${e}\n`); }
	}

	// Sort by score descending, truncate
	results.sort((a, b) => b.score - a.score);
	return results.slice(0, limit);
}

function fuzzySearchKnowledge(
	store: Store,
	queryWords: string[],
	limit: number,
	exclude: KnowledgeRow[],
): KnowledgeRow[] {
	if (limit <= 0) return [];
	const excludeIds = new Set(exclude.map((d) => d.id));

	const rows = store.db
		.prepare(`
    SELECT id, project_id, file_path, content_hash, title, content, sub_type,
           branch, created_at, updated_at, hit_count, last_accessed, enabled
    FROM knowledge_index WHERE enabled = 1 LIMIT 500
  `)
		.all() as RawKnowledgeRow[];

	const docs: KnowledgeRow[] = [];
	for (const r of rows) {
		if (excludeIds.has(r.id)) continue;

		const targetWords = r.title.toLowerCase().split(/\s+/);
		for (const qw of queryWords) {
			let matched = false;
			for (const tw of targetWords) {
				if (fuzzyMatch(qw, tw)) {
					matched = true;
					break;
				}
			}
			if (matched) {
				docs.push(mapRow(r));
				if (docs.length >= limit) return docs;
				break;
			}
		}
	}
	return docs;
}

export function detectKnowledgeConflicts(
	store: Store,
	threshold = 0.7,
	limit = 1000,
): KnowledgeConflict[] {
	const pairs = pairwiseSimilarity(store, { limit, minScore: threshold });

	const conflicts: KnowledgeConflict[] = [];
	for (const pair of pairs) {
		if (pair.score >= threshold) {
			conflicts.push({
				a: { id: pair.idA } as KnowledgeRow,
				b: { id: pair.idB } as KnowledgeRow,
				similarity: pair.score,
				type: "potential_duplicate",
			});
		}
	}

	if (conflicts.length > 0) {
		const allIds = conflicts.flatMap((c) => [c.a.id, c.b.id]);
		const hydrated = getKnowledgeByIDs(store, allIds);
		const docMap = new Map(hydrated.map((d) => [d.id, d]));
		for (const conflict of conflicts) {
			const a = docMap.get(conflict.a.id);
			const b = docMap.get(conflict.b.id);
			if (a) conflict.a = a;
			if (b) conflict.b = b;
			conflict.type = classifyConflict(conflict.a.content, conflict.b.content);
		}
	}

	conflicts.sort((a, b) => b.similarity - a.similarity);
	return conflicts;
}

export function expandAliases(store: Store, terms: string[]): string[] {
	if (terms.length === 0) return [];

	const seen = new Set(terms.map((t) => t.toLowerCase()));

	for (const t of terms) {
		const lower = t.toLowerCase();

		const aliasRows = store.db
			.prepare("SELECT alias FROM tag_aliases WHERE LOWER(tag) = ?")
			.all(lower) as Array<{ alias: string }>;
		for (const r of aliasRows) seen.add(r.alias.toLowerCase());

		const tagRows = store.db
			.prepare("SELECT tag FROM tag_aliases WHERE LOWER(alias) = ?")
			.all(lower) as Array<{ tag: string }>;
		for (const r of tagRows) seen.add(r.tag.toLowerCase());
	}

	return [...seen];
}

function sanitizeFTSTerm(term: string): string {
	return term.replace(/["*^{}]/g, "");
}

const CONTRADICTION_PAIRS: [string, string][] = [
	["always", "never"],
	["must", "must not"],
	["use", "avoid"],
	["enable", "disable"],
	["allow", "deny"],
	["required", "optional"],
	["do", "don't"],
	["add", "remove"],
	["include", "exclude"],
];

export function classifyConflict(
	contentA: string,
	contentB: string,
): "potential_duplicate" | "potential_contradiction" {
	const lowerA = contentA.toLowerCase();
	const lowerB = contentB.toLowerCase();

	for (const [w0, w1] of CONTRADICTION_PAIRS) {
		const aHas0 = lowerA.includes(w0);
		const aHas1 = lowerA.includes(w1);
		const bHas0 = lowerB.includes(w0);
		const bHas1 = lowerB.includes(w1);

		if ((aHas0 && bHas1 && !aHas1) || (aHas1 && bHas0 && !bHas1)) {
			return "potential_contradiction";
		}
	}
	return "potential_duplicate";
}

export function levenshtein(a: string, b: string): number {
	const ra = [...a];
	const rb = [...b];
	const la = ra.length;
	const lb = rb.length;

	if (la === 0) return lb;
	if (lb === 0) return la;

	let prev = Array.from({ length: lb + 1 }, (_, j) => j);

	for (let i = 1; i <= la; i++) {
		const curr = new Array<number>(lb + 1);
		curr[0] = i;
		for (let j = 1; j <= lb; j++) {
			const cost = ra[i - 1] === rb[j - 1] ? 0 : 1;
			curr[j] = Math.min(curr[j - 1]! + 1, prev[j]! + 1, prev[j - 1]! + cost);
		}
		prev = curr;
	}
	return prev[lb]!;
}

export function fuzzyMatch(query: string, target: string): boolean {
	const qLen = [...query].length;
	if (qLen < 3) return false;
	let maxDist = Math.min(2, Math.floor(qLen / 3));
	if (maxDist === 0) maxDist = 1;
	return levenshtein(query.toLowerCase(), target.toLowerCase()) <= maxDist;
}
